### PR TITLE
feat: load read buffer on demand

### LINKED ISSUE #5032: Lazy parquet chunk => RB chunk conversion
# Why
Certain parts of our query stack assume that gathering chunks does not pull data. The querier somewhat violates that pattern by instancing RBs for every requested parquet file. We should delay this conversion as long as possible.

The assumption of the query stack is both practical (i.e. it simplifies the implementation of some of our InfluxRPC routines) as well as necessary (imagine a `SELECT * from table limit 1`). #4996 improved the situation a bit but does not cover all use cases (e.g. https://github.com/influxdata/conductor/issues/1012 )

# What
The implementation of `QuerierParquetChunk` and `QuerierRBChunk` are very close.

| Method                                  | identical | RB :arrow_up: quality | RB faster |
| --------------------------------------- | --------- | --------------------- | --------- |
| `QueryChunkMeta::summary`               |           | X[^su]                |           |
| `QueryChunkMeta::schema`                | X         |                       |           |
| `QueryChunkMeta::partition_sort_key`    | X         |                       |           |
| `QueryChunkMeta::partition_id`          | X         |                       |           |
| `QueryChunkMeta::sort_key`              | X         |                       |           |
| `QueryChunkMeta::delete_predicates`     | X         |                       |           |
| `QueryChunkMeta::timestamp_min_max`     | X         |                       |           |
| `QueryChunk::id`                        | X         |                       |           |
| `QueryChunk::table_name`                | X         |                       |           |
| `QueryChunk::may_contain_pk_duplicates` | X         |                       |           |
| `QueryChunk::column_names`              |           | X                     |           |
| `QueryChunk::column_values`             |           | X                     |           |
| `QueryChunk::read_filter`               |           |                       | X         |
| `QueryChunk::chunk_type`                | X[^ct]    |                       |           |
| `QueryChunk::order`                     | X         |                       |           |

[^su]: as of #5030 this also improves `QueryChunk::apply_predicate_to_metadata`
[^ct]: both return a constant value, but the value is different

All methods are sync, except for `QueryChunk::read_filter` which contains a `SendableRecordBatchStream` and therefore is somewhat async. Therefore this is the only method that would allow lazy loading of data.

# How
As shown in the section above, there are only a few methods that actually differentiate the RB version from the parquet version. Hence, the following schema should be implemented:

- **chunk creation:** when the chunk is created, the cache should be probed for the RB. If the RB is present, it is used. If it is not present, it MUST NOT be loaded at this point (this is different from a normal cache request). The reasoning is that if we already have the RB available, we can easily use it to improve our statistics.
- **metadata request:** use RB or parquet depending on what was used during chunk creation. If we only have parquet data, we have to live with the lower-quality statistics.
- **data request (`read_filter`):** Since reading from parquet requires large data transfers anyways[^parquet], hydrate the parquet into an RB, e.g. perform an ordinary RB cache request. This will automatically also improve subsequent metadata requests.

[^parquet]: At least currently, this might change in the future. The technique proposed in this ticket is future-proof, since we could still decide if we want to hydrate the entire parquet into an RB (e.g. based on query stats or on how much data is requested).

### PR DESCRIPTION
**Pull request by crepererum**
Closes #5032.

### CODE REVIEW

File: querier/src/chunk/mod.rs [hunk 1]
```diff
@@ -452,62 +530,235 @@ pub mod tests {
...
+
+        // check state
+        assert_eq!(chunk.chunk_type(), "parquet");
 
         // measure catalog access
-        let catalog_metrics1 = get_catalog_access_metrics(&catalog.metric_registry());
+        let catalog_metrics1 = test_data.get_catalog_access_metrics();
 
         // check chunk schema
+        assert_schema(&chunk);
```
Comment (alamb) [inline querier/src/chunk/mod.rs:592]: this is a very nice test structure

File: querier/src/chunk/query_access.rs [hunk 2]
```diff
@@ -400,3 +455,40 @@ impl futures::Stream for ReadFilterResultsStream {
 
     // TODO is there a useful size_hint to pass?
 }
+
+enum OwnedSelection {
+    All,
+    Some(Vec<String>),
+}
+
+enum HalfOwnedSelection<'a> {
```
Comment (alamb) [inline querier/src/chunk/query_access.rs:464]: lol

File: querier/src/chunk/query_access.rs [hunk 3]
```diff
@@ -400,3 +455,40 @@ impl futures::Stream for ReadFilterResultsStream {
 
     // TODO is there a useful size_hint to pass?
 }
+
+enum OwnedSelection {
```
Comment (alamb) [inline querier/src/chunk/query_access.rs:459]: eventually this might be nice to put into `schema` crate (along side `Selection`)

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
This is really nice

File: parquet_file/src/storage.rs [hunk 4]
```diff
@@ -226,35 +223,6 @@ impl ParquetStorage {
     }
 }
 
-/// Return indices of the schema's fields of the selection columns
-fn column_indices(selection: Selection<'_>, schema: SchemaRef) -> Vec<usize> {
```
[the remark this answers is not in the document]
Comment (crepererum) [reply parquet_file/src/storage.rs:230]: yes, see first commit

File: querier/src/chunk/mod.rs [hunk 5]
```diff
@@ -125,6 +127,66 @@ impl ChunkStage {
...
+    fn from(rb_chunk: Arc<RBChunk>) -> Self {
+        let table_summary = Arc::new(rb_chunk.table_summary());
+        Self::ReadBuffer {
+            rb_chunk,
+            table_summary,
+        }
+    }
+}
+
+/// Controls if/how data for a [`QuerierChunk`] is loaded
```
[the remark this answers is not in the document]
Comment (alamb) [reply querier/src/chunk/mod.rs:172]: Sounds reasonable to me

File: querier/src/chunk/query_access.rs (hunk 3 above)
Comment (crepererum) [reply querier/src/chunk/query_access.rs]: done

**Pull request merged**